py/objdict.c: Fix for JSON dump of dicts that have numeric keys

MrSurly opened the pull request:
Per https://github.com/micropython/micropython/issues/4790

This implementation is likely too naive for @dpgeorge, but this is a first attempt.

MrSurly committed:
py/objdict.c: Fix for https://github.com/micropython/micropython/issues/4790

dpgeorge commented on py/objdict.c:
```diff
@@ -70,7 +74,14 @@ STATIC void dict_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_
             mp_print_str(print, ", ");
         }
         first = false;
+	bool quote = quote_nums && (mp_obj_is_int(next->key) || mp_obj_is_float(next->key));
```
To cover the case of True/False/None it's probably better to have this as a check for strings, and if not a string then add quotes.  That would also be more efficient, only a single check.

MrSurly replied:
Pushed update, per your suggestions.

MrSurly committed:
py/objdict.c: Any non-string value is now quoted, per suggestion from @dpgeorge

MrSurly committed:
py/objdict.c: Any non-string value is now quoted, per suggestion from @dpgeorge

dpgeorge commented:
Thanks for that.

I'll have a look if there's any way to reduce code size.

Also, eventually it'll need some tests to cover the cases that it fixes.

MrSurly committed:
Merge branch 'master' into json_dumps_key_fix

MrSurly committed:
Merge branch 'master' into json_dumps_key_fix

MrSurly commented:
> Also, eventually it'll need some tests to cover the cases that it fixes.

Is it as simple as copying `tests/extmod/ujson_dump.py` and modifying it?  

I've looked at these, and it's not clear how the testing system "knows" if a particular test passed.  Are the scripts supposed to raise an exception when a test fails?

stinos commented:
The test files are ran using the MicroPython executable, print some output, and that output is compared against the epected output. Which is either read from a <testfile>.py.exp file or, in absence of that .exp file, by running the testfile with Cpython as well and using that output as expected output.

In this case you don't really need to copy ujson_dumps.py (and it's _loads counterpart), but just add extra lines to it covering the new code i.e. loading/dumping dicts with numeric keys hereby proving the code works as epected.

MrSurly commented:
@stinos 

I'll add tests there, then.  Thanks.

MrSurly committed:
tests/extmod/ujson.dump.py: Test dumping objects with numerical keys

MrSurly committed:
tests/extmod/ujson.dump.py: Test dumping objects with numerical keys

stinos commented:
Should have highlighted this, but I think the tests should go in `ujson_dumps` not `ujson_dump`: the latter is more about testing whether `dump` works whereas the former tests all supported types.

MrSurly committed:
tests/extmod/ujson.dumps.py: Test dumping objects with numerical keys

MrSurly commented:
> Should have highlighted this, but I think the tests should go in `ujson_dumps` not `ujson_dump`: the latter is more about testing whether `dump` works whereas the former tests all supported types.

Done.

dpgeorge commented:
@MrSurly re this comment: https://github.com/micropython/micropython/issues/4790#issuecomment-513119228 ; could you give some insight as to why it's not feasible/possible/convenient to do (which is already supported):
```python
d = {"1":1, "1.2":1.2, "true":True}
json.dumps(d)
```

dpgeorge commented on tests/extmod/ujson_dumps.py:
```diff
@@ -26,3 +26,5 @@
 print(json.dumps({"a":(2,[3,None])}))
 print(json.dumps('"quoted"'))
 print(json.dumps('space\n\r\tspace'))
+print(json.dumps({1:2, 3:4, 5.5:6}))
```
Please just put one element in the dict, otherwise the test may fail if the order is different when printing.

Also, float tests need to go in the separate file `ujson_dumps_float.py`.

MrSurly replied:
Understood why this is important (undefined key order), but doesn't that mean we're missing coverage of multi-key objects?

In any case, I'll change it up.

MrSurly replied:
Updated

dpgeorge replied:
> but doesn't that mean we're missing coverage of multi-key objects?

True.  A way to fix that would be to sort the resulting string, eg `print(sorted(json.dumps({1:2, 3:4})))`, but that's no an ideal solution.

MrSurly committed:
tests/extmod/ujson_dumps.py
tests/extmod/ujson_dumps_float.py

Updated tests per https://github.com/micropython/micropython/pull/4926#pullrequestreview-266465534

MrSurly commented:
> @MrSurly re this comment: [#4790 (comment)](https://github.com/micropython/micropython/issues/4790#issuecomment-513119228) ; could you give some insight as to why it's not feasible/possible/convenient to do (which is already supported):

Reply here: https://github.com/micropython/micropython/issues/4790#issuecomment-515152384

Text:

> Firstly, `json.dump` and `json.dumps` should _only produce valid JSON_, or throw an exception when something can't be properly serialized to JSON.
> 
> For my use case, I have abstracted classes that return a Python data structures, that I then serialize (using the `json` class) for transmission via MQTT. With `ujson` in it's present state, then I'd either have to modify underlying classes to be aware of the limitation with `ujson`, or modify the layers above to do a deep traversal of the data structure to "fix" it so that `ujson` behaves correctly.

dpgeorge commented:
Ok, thanks for the explanation.  I agree the case should at least be caught, and quoting the keys is easy once the case is tested for.

Merged in 01054f20925b92229402750979864f8d38a76916 with some minor changes and additional tests.
